File: packages/adapter-acp/src/agents.ts
import { existsSync, rmSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'
import type { Account, Model, ProviderSetup } from '@harness/contracts'
import { isInstalled, killTree, spawnCli } from '@harness/proc'
import { z } from 'zod'
import { propertiesWhen } from './properties-when.js'

const KimiModelsSchema = z.object({
  models: z
    .record(
      z.string(),
      z.object({
        displayName: z.string().optional(),
        supportEfforts: z.array(z.string()).optional(),
        defaultEffort: z.string().optional(),
      }),
    )
    .default({}),
})

/**
 * Agents we know how to launch in ACP mode.
 *
 * This is a list of launch commands, not a list of integrations. Every entry
 * goes through the same adapter — that is the entire argument for supporting
 * ACP at all. Adding an agent is a row here.
 *
 * `verified` marks the ones whose frames were actually captured and read while
 * writing the adapter. The rest are launched on the strength of their own
 * documentation, so they are offered but labelled honestly rather than
 * presented as tested.
 */

export type AcpAgentSpec = {
  id: string
  name: string
  /** Executable, resolved on PATH. May be an npm shim on Windows. */
  command: string
  args: string[]
  /** True when frames from this agent were captured and read by us. */
  verified: boolean
  /** Wire version captured while verifying this agent. */
  supportedVersion?: string
  /** Shown when the binary is missing, so the user knows what to install. */
  install?: string
  setup: ProviderSetup
  /** Why sign-in or use is impaired right now, in words shown to the user. */
  problem?: string
  /** CLI flag used to select a model before the ACP handshake. */
  modelArg?: string
  /** ACP config option used to select a model after creating a session. */
  modelConfigId?: string
  /**
   * No longer offered anywhere in the product. The spec itself must remain:
   * resuming a persisted session still needs the launch command, and deleting
   * the row would turn old threads into a crash instead of a session.
   */
  retired?: boolean
  /**
   * File whose existence marks a completed provider login, relative to the
   * user's home directory. Existence only — the content is the provider's
   * credential and is never read. Absent when the CLI gives us no way to
   * tell, in which case sign-in state stays unknown.
   */
  credentialProbe?: string
}

export const ACP_AGENTS: AcpAgentSpec[] = [
  {
    id: 'gemini',
    name: 'Gemini CLI',
    command: 'gemini',
    args: ['--experimental-acp'],
    verified: true,
    modelArg: '--model',
    install: 'npm i -g @google/gemini-cli',
    setup: {
      installUrl: 'https://github.com/google-gemini/gemini-cli/blob/main/docs/get-started/index.md',
      installCommand: 'npm install -g @google/gemini-cli',
      login: 'provider',
    },
    // Google discontinued Gemini Code Assist for individuals on 2026-06-18
    // and Antigravity is the direct provider that replaced it, so the row is
    // retired from every listing (decided 2026-08-07). Existing gemini
    // threads still resume through this spec.
    retired: true,
    problem:
      'Google ended individual sign-in (June 2026) — use an organization account or set GEMINI_API_KEY.',
  },
  {
    id: 'kimi',
    name: 'Kimi CLI',
    command: 'kimi',
    args: ['acp'],
    verified: true,
    supportedVersion: '0.29',
    modelConfigId: 'model',
    // kimi-code 0.29.1 has no auth-status command (`kimi --help` offers only
    // login/provider/doctor), and the ACP initialize response advertises auth
    // *methods* regardless of state. Its config.toml keys `oauth/kimi-code`
    // into this credentials store, so the file's existence is the one signal
    // a completed device-code login leaves behind. Captured 2026-08-07.
    credentialProbe: join('.kimi-code', 'credentials', 'kimi-code.json'),
    install: 'npm install -g @moonshot-ai/kimi-code',
    setup: {
      installUrl: 'https://moonshotai.github.io/kimi-code/en/guides/getting-started.html',
      installCommand: 'npm install -g @moonshot-ai/kimi-code',
      login: 'provider',
    },
  },
  {
    id: 'qwen',
    name: 'Qwen Code',
    command: 'qwen',
    args: ['--experimental-acp'],
    // A Gemini CLI fork, so the same flag and very likely the same frames.
    // Not claimed as verified until someone has actually run it.
    verified: false,
    modelArg: '--model',
    install: 'npm i -g @qwen-code/qwen-code',
    setup: {
      installUrl: 'https://qwenlm.github.io/qwen-code-docs/en/users/quickstart/',
      installCommand: 'npm install -g @qwen-code/qwen-code@latest',
      login: 'provider',
    },
    // Dropped from the roster on Leon's call (2026-08-08); existing qwen
    // threads still resume through this spec, same as gemini.
    retired: true,
  },
]

/** The agents offered in listings — retirement hides an agent everywhere new. */
export const LISTED_AGENTS: AcpAgentSpec[] = ACP_AGENTS.filter((agent) => !agent.retired)

/** Resolves retired agents too: resume must outlive the listing. */
export function findAgentSpec(id: string): AcpAgentSpec | undefined {
  return ACP_AGENTS.find((agent) => agent.id === id)
}

export async function detectAgents(): Promise<Array<AcpAgentSpec & { installed: boolean }>> {
  return Promise.all(
    LISTED_AGENTS.map(async (agent) => ({ ...agent, installed: await isInstalled(agent.command) })),
  )
}

/**
 * Sign-in state as far as the agent's CLI lets us observe it. Agents declare
 * the file a completed login leaves behind (`credentialProbe`); we check that
 * it exists and nothing more — reading it would cross the credential
 * boundary in rules/security.md. Without a probe the state is unknown and
 * reported as signed-out, which keeps the sign-in flow reachable.
 */
export function acpAccount(agentId: string, home = homedir()): Account {
  const spec = findAgentSpec(agentId)
  if (!spec?.credentialProbe) return { signedIn: false }
  return { signedIn: existsSync(join(home, spec.credentialProbe)) }
}

/**
 * Sign out by removing the credential the login left behind — the CLIs
 * offer no logout command, and deleting the file is exactly what one would
 * do. The file is removed, never read. Throws for agents without a probe so
 * the UI knows not to offer the action.
 */
export function acpSignOut(agentId: string, home = homedir()): void {
  const spec = findAgentSpec(agentId)
  if (!spec?.credentialProbe) {
    throw new Error(`${spec?.name ?? agentId} does not support signing out from here`)
  }
  rmSync(join(home, spec.credentialProbe), { force: true })
}

/**
 * Model choices exposed by the agent's own stable surface.
 *
 * The Gemini ids are the concrete names `--model` accepts, taken from
 * gemini-cli-core's VALID_GEMINI_MODELS (verified against gemini-cli 0.21.x),
 * plus the `auto` routing alias. Concrete names are what people recognise from
 * Google's own docs; `auto` stays the default because routing is the CLI's
 * recommended mode.
 */
export async function discoverAgentModels(agentId: string): Promise<Model[]> {
  if (agentId === 'gemini') {
    return [
      model('gemini-3-pro-preview', 'Gemini 3 Pro (Preview)', 'Most capable Gemini model', true),
      model('gemini-3-flash-preview', 'Gemini 3 Flash (Preview)', 'Fast Gemini 3 model'),
      model('gemini-2.5-pro', 'Gemini 2.5 Pro', 'Stable Pro model'),
      model('gemini-2.5-flash', 'Gemini 2.5 Flash', 'Stable fast model'),
      model('gemini-2.5-flash-lite', 'Gemini 2.5 Flash Lite', 'Lightest and cheapest model'),
    ]
  }
  if (agentId !== 'kimi') return []
  let models: Model[]
  try {
    // Kept for Kimi 0.x, whose CLI exposes this catalog. Kimi 1.5 removed the
    // command and its ACP session currently exposes no config options, so that
    // version falls back to the provider default instead of showing controls
    // the wire cannot apply.
    models = parseKimiModels(await captureCli('kimi', ['provider', 'list', '--json']))
  } catch {
    return []
  }
  // The legacy catalog can name graded efforts, but the ACP version paired
  // with it only drives model selection. Keep the parser accurate while the
  // selectable rows report only controls this adapter can actually apply.
  return models.map((model) => {
    const selectable = { ...model, reasoningEfforts: [] }
    delete selectable.defaultReasoningEffort
    return selectable
  })
}

function model(id: string, displayName: string, description: string, isDefault = false): Model {
  return { id, displayName, description, isDefault, reasoningEfforts: [], serviceTiers: [] }
}

export function parseKimiModels(output: string): Model[] {
  const parsed = KimiModelsSchema.parse(JSON.parse(output))
  return Object.entries(parsed.models).map(([id, details], index) => {
    return {
      id,
      displayName: details.displayName ?? id,
      isDefault: index === 0,
      reasoningEfforts: details.supportEfforts ?? [],
      ...propertiesWhen(details.defaultEffort, (defaultReasoningEffort) => ({
        defaultReasoningEffort,
      })),
      serviceTiers: [],
    }
  })
}

function captureCli(command: string, args: string[], timeoutMs = 5000): Promise<string> {
  return new Promise((resolve, reject) => {
    const child = spawnCli(command, args)
    let output = ''
    let settled = false
    const finish = (error?: Error) => {
      if (settled) return
      settled = true
      clearTimeout(timer)
      error ? reject(error) : resolve(output)
    }
    const timer = setTimeout(() => {
      killTree(child)
      finish(new Error(`${command} model discovery timed out`))
    }, timeoutMs)
    child.stdout.setEncoding('utf8')
    child.stdout.on('data', (chunk: string) => (output += chunk))
    child.on('error', (error) => finish(error))
    child.on('close', (code) =>
      finish(code === 0 ? undefined : new Error(`${command} model discovery failed`)),
    )
  })
}

File: crates/harness-adapter-acp/src/agents.rs
use harness_agent::{AgentError, AgentResult};
use harness_protocol::{Account, Model};
use serde::Deserialize;
use serde_json::Value;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct AcpAgentSpec {
    pub id: &'static str,
    pub name: &'static str,
    pub command: &'static str,
    pub args: &'static [&'static str],
    pub verified: bool,
    pub supported_version: Option<&'static str>,
    pub model_arg: Option<&'static str>,
    pub model_config_id: Option<&'static str>,
    pub retired: bool,
    pub credential_probe: Option<&'static [&'static str]>,
}

const ACP_AGENTS: &[AcpAgentSpec] = &[
    AcpAgentSpec {
        id: "gemini",
        name: "Gemini CLI",
        command: "gemini",
        args: &["--experimental-acp"],
        verified: true,
        supported_version: None,
        model_arg: Some("--model"),
        model_config_id: None,
        retired: true,
        credential_probe: None,
    },
    AcpAgentSpec {
        id: "kimi",
        name: "Kimi CLI",
        command: "kimi",
        args: &["acp"],
        verified: true,
        supported_version: Some("0.29"),
        model_arg: None,
        model_config_id: Some("model"),
        retired: false,
        credential_probe: Some(&[".kimi-code", "credentials", "kimi-code.json"]),
    },
    AcpAgentSpec {
        id: "qwen",
        name: "Qwen Code",
        command: "qwen",
        args: &["--experimental-acp"],
        verified: false,
        supported_version: None,
        model_arg: Some("--model"),
        model_config_id: None,
        retired: true,
        credential_probe: None,
    },
];

pub fn find_agent_spec(id: &str) -> Option<&'static AcpAgentSpec> {
    ACP_AGENTS.iter().find(|agent| agent.id == id)
}

pub fn listed_agent_specs() -> impl Iterator<Item = &'static AcpAgentSpec> {
    ACP_AGENTS.iter().filter(|agent| !agent.retired)
}

pub fn acp_account(agent_id: &str, home: &Path) -> Account {
    let signed_in = find_agent_spec(agent_id)
        .and_then(|spec| spec.credential_probe)
        .is_some_and(|probe| credential_path(home, probe).is_file());
    Account {
        signed_in,
        email: None,
        plan: None,
    }
}

pub fn acp_sign_out(agent_id: &str, home: &Path) -> AgentResult<()> {
    let spec = find_agent_spec(agent_id);
    let probe = spec
        .and_then(|spec| spec.credential_probe)
        .ok_or(AgentError::Unsupported("signing out from this ACP agent"))?;
    match std::fs::remove_file(credential_path(home, probe)) {
        Ok(()) => Ok(()),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(_) => Err(AgentError::Failed(format!(
            "{} could not sign out",
            spec.map_or(agent_id, |spec| spec.name)
        ))),
    }
}

fn credential_path(home: &Path, components: &[&str]) -> PathBuf {
    components
        .iter()
        .fold(home.to_path_buf(), |path, component| path.join(component))
}

pub fn gemini_models() -> Vec<Model> {
    vec![
        model(
            "gemini-3-pro-preview",
            "Gemini 3 Pro (Preview)",
            "Most capable Gemini model",
            true,
        ),
        model(
            "gemini-3-flash-preview",
            "Gemini 3 Flash (Preview)",
            "Fast Gemini 3 model",
            false,
        ),
        model(
            "gemini-2.5-pro",
            "Gemini 2.5 Pro",
            "Stable Pro model",
            false,
        ),
        model(
            "gemini-2.5-flash",
            "Gemini 2.5 Flash",
            "Stable fast model",
            false,
        ),
        model(
            "gemini-2.5-flash-lite",
            "Gemini 2.5 Flash Lite",
            "Lightest and cheapest model",
            false,
        ),
    ]
}

pub fn parse_kimi_models(output: &str) -> AgentResult<Vec<Model>> {
    #[derive(Deserialize)]
    struct Wire {
        #[serde(default, deserialize_with = "ordered_values")]
        models: Vec<(String, Value)>,
    }

    let wire: Wire = serde_json::from_str(output)
        .map_err(|_| AgentError::Failed("Kimi model discovery returned invalid JSON".into()))?;
    Ok(wire
        .models
        .into_iter()
        .enumerate()
        .map(|(index, (id, details))| {
            let display_name = details
                .get("displayName")
                .and_then(Value::as_str)
                .unwrap_or(&id)
                .to_owned();
            let reasoning_efforts = details
                .get("supportEfforts")
                .and_then(Value::as_array)
                .into_iter()
                .flatten()
                .filter_map(Value::as_str)
                .map(str::to_owned)
                .collect();
            let default_reasoning_effort = details
                .get("defaultEffort")
                .and_then(Value::as_str)
                .map(str::to_owned);
            Model {
                id,
                display_name,
                description: None,
                is_default: index == 0,
                reasoning_efforts,
                default_reasoning_effort,
                service_tiers: Vec::new(),
                default_service_tier: None,
            }
        })
        .collect())
}

fn ordered_values<'de, D>(deserializer: D) -> Result<Vec<(String, Value)>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    struct Visitor;

    impl<'de> serde::de::Visitor<'de> for Visitor {
        type Value = Vec<(String, Value)>;

        fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            formatter.write_str("a Kimi model object")
        }

        fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::MapAccess<'de>,
        {
            let mut values = Vec::new();
            while let Some(entry) = map.next_entry()? {
                values.push(entry);
            }
            Ok(values)
        }
    }

    deserializer.deserialize_map(Visitor)
}

fn model(id: &str, display_name: &str, description: &str, is_default: bool) -> Model {
    Model {
        id: id.into(),
        display_name: display_name.into(),
        description: Some(description.into()),
        is_default,
        reasoning_efforts: Vec::new(),
        default_reasoning_effort: None,
        service_tiers: Vec::new(),
        default_service_tier: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn records_the_captured_agent_launch_contracts() {
        let kimi = find_agent_spec("kimi").unwrap();
        assert_eq!(kimi.command, "kimi");
        assert_eq!(kimi.args, ["acp"]);
        assert!(kimi.verified);
        assert_eq!(kimi.supported_version, Some("0.29"));
        assert!(!kimi.retired);
        assert_eq!(
            listed_agent_specs()
                .map(|agent| agent.id)
                .collect::<Vec<_>>(),
            ["kimi"]
        );
        assert!(find_agent_spec("gemini").unwrap().retired);
        assert!(find_agent_spec("qwen").unwrap().retired);
        assert!(find_agent_spec("missing").is_none());
    }

    #[test]
    fn probes_and_removes_only_kimis_credential_file() {
        let home = tempfile::tempdir().unwrap();
        assert!(!acp_account("kimi", home.path()).signed_in);
        let credentials = home.path().join(".kimi-code").join("credentials");
        std::fs::create_dir_all(&credentials).unwrap();
        let credential = credentials.join("kimi-code.json");
        std::fs::write(&credential, b"{}").unwrap();
        let sibling = credentials.join("keep.json");
        std::fs::write(&sibling, b"{}").unwrap();
        assert!(acp_account("kimi", home.path()).signed_in);
        acp_sign_out("kimi", home.path()).unwrap();
        assert!(!credential.exists());
        assert!(sibling.exists());
        assert!(!acp_account("qwen", home.path()).signed_in);
        assert!(acp_sign_out("qwen", home.path()).is_err());
    }

    #[test]
    fn offers_the_concrete_gemini_model_names() {
        let models = gemini_models();
        assert_eq!(
            models
                .iter()
                .map(|model| model.id.as_str())
                .collect::<Vec<_>>(),
            [
                "gemini-3-pro-preview",
                "gemini-3-flash-preview",
                "gemini-2.5-pro",
                "gemini-2.5-flash",
                "gemini-2.5-flash-lite",
            ]
        );
        assert!(models[0].is_default);
        assert_eq!(models.iter().filter(|model| model.is_default).count(), 1);
    }

    #[test]
    fn normalizes_kimi_models_without_reordering_the_wire() {
        let models = parse_kimi_models(
            r#"{
                "models": {
                    "z-model": {
                        "displayName": "Kimi Z",
                        "supportEfforts": ["low", 7, "high"],
                        "defaultEffort": "high"
                    },
                    "a-model": { "displayName": "Kimi A" }
                }
            }"#,
        )
        .unwrap();
        assert_eq!(models[0].id, "z-model");
        assert_eq!(models[0].display_name, "Kimi Z");
        assert!(models[0].is_default);
        assert_eq!(models[0].reasoning_efforts, ["low", "high"]);
        assert_eq!(models[0].default_reasoning_effort.as_deref(), Some("high"));
        assert_eq!(models[1].id, "a-model");
        assert!(!models[1].is_default);
    }
}
